pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -569,13 +576,44 @@ private SegmentPartitionMetadata getPartitionMetadataFromTableConfig(TableConfig
     for (Map.Entry<String, ColumnPartitionConfig> entry : partitionConfig.getColumnPartitionMap().entrySet()) {
       String columnName = entry.getKey();
       ColumnPartitionConfig columnPartitionConfig = entry.getValue();
+
+      // NOTE: Here we compare the number of partitions from the config and the stream, and log a warning when they
+      //       don't match, but use the one from the stream. The mismatch could happen when the stream partitions are
+      //       changed, but the table config has not been updated to reflect the change. In such case, picking the
+      //       number of partitions from the stream can keep the segment properly partitioned as long as the partition
+      //       function is not changed.
+      if (columnPartitionConfig.getNumPartitions() != numPartitions) {
```

Comment:
For realtime table, maybe we should not have the num partitions in the tableconfig. It can be set to -1, to indicate that it needs to come from the stream.

Reply:
Good point, or we just don't read it. While we can still use it when the server failed to get the numPartitions from the stream (the existing behavior).